csi/controller_test.go

```diff
@@ -911,3 +911,222 @@ func TestControllerListVolumes(t *testing.T) {
+	name := "myvol"
+	size := uint64(1234)
+	req := &csi.CreateVolumeRequest{
+		Version: &csi.Version{},
+		Name:    name,
+		VolumeCapabilities: []*csi.VolumeCapability{
+			&csi.VolumeCapability{},
+		},
+		CapacityRange: &csi.CapacityRange{
+			RequiredBytes: size,
+		},
+		Parameters: map[string]string{
+			api.SpecParent: mockParentID,
+		},
+	}
+
+	// Setup mock functions
+	id := "myid"
+	gomock.InOrder(
+		s.MockDriver().
+			EXPECT().
+			Inspect([]string{name}).
+			Return(nil, fmt.Errorf("not found")).
+			Times(1),
+
+		s.MockDriver().
+			EXPECT().
+			Enumerate(&api.VolumeLocator{Name: name}, nil).
+			Return(nil, fmt.Errorf("not found")).
+			Times(1),
+
+		s.MockDriver().
+			EXPECT().
+			Inspect([]string{mockParentID}).
+			Return([]*api.Volume{
+				&api.Volume{
+					Id: mockParentID,
+				},
+			}, nil).
+			Times(1),
+
+		s.MockDriver().
+			EXPECT().
+			Snapshot(mockParentID, false, &api.VolumeLocator{
+				Name: name,
+			}).
+			Return(id, nil).
+			Times(1),
+
+		s.MockDriver().
+			EXPECT().
+			Inspect([]string{id}).
+			Return([]*api.Volume{
+				&api.Volume{
+					Id: id,
+					Locator: &api.VolumeLocator{
+						Name: name,
+					},
+					Spec: &api.VolumeSpec{
+						Size: size,
+					},
+				},
+			}, nil).
+			Times(1),
+	)
+
+	r, err := c.CreateVolume(context.Background(), req)
+	assert.Nil(t, err)
+	assert.NotNil(t, r)
+	volumeInfo := r.GetResult().GetVolumeInfo()
+
+	assert.Equal(t, id, volumeInfo.GetId())
+	assert.Equal(t, size, volumeInfo.GetCapacityBytes())
+}
+
+func TestControllerDeleteVolumeInvalidArguments(t *testing.T) {
+	// Create server and client connection
+	s := newTestServer(t)
+	defer s.Stop()
+	c := csi.NewControllerClient(s.Conn())
+
+	// No version
+	req := &csi.DeleteVolumeRequest{}
+	_, err := c.DeleteVolume(context.Background(), req)
+	assert.NotNil(t, err)
+	serverError, ok := status.FromError(err)
+	assert.True(t, ok)
+	assert.Equal(t, serverError.Code(), codes.InvalidArgument)
+	assert.Contains(t, serverError.Message(), "Version")
+
+	// No id
+	req.Version = &csi.Version{}
+	_, err = c.DeleteVolume(context.Background(), req)
+	assert.NotNil(t, err)
+	serverError, ok = status.FromError(err)
+	assert.True(t, ok)
+	assert.Equal(t, serverError.Code(), codes.InvalidArgument)
+	assert.Contains(t, serverError.Message(), "Volume id")
+}
+
+func TestControllerDeleteVolumeError(t *testing.T) {
+	// Create server and client connection
+	s := newTestServer(t)
+	defer s.Stop()
+	c := csi.NewControllerClient(s.Conn())
+
+	// No version
+	myid := "myid"
+	req := &csi.DeleteVolumeRequest{
+		Version:  &csi.Version{},
+		VolumeId: myid,
+	}
+
+	// Setup mock
+	s.MockDriver().EXPECT().Delete(myid).Return(fmt.Errorf("TEST")).Times(1)
+
+	_, err := c.DeleteVolume(context.Background(), req)
+	assert.NotNil(t, err)
+	serverError, ok := status.FromError(err)
+	assert.True(t, ok)
+	assert.Equal(t, serverError.Code(), codes.Internal)
+	assert.Contains(t, serverError.Message(), "Unable to delete")
```

Comment:
nit: should we check for "TEST" to be sure the error is coming because of the delete call failing? in case, we are using "unable to delete" in multiple places.

Reply:
👍